src/nccl_ofi_net.c

```diff
@@ -158,85 +158,204 @@ void nccl_net_ofi_free_info_list(struct fi_info *info_list)
+	}
+
+	/* Initialize mutex */
+	ret = pthread_mutex_init(&id_pool->lock, NULL);
+	if (OFI_UNLIKELY(ret)) {
+		NCCL_OFI_WARN("Unable to initialize mutex");
+		free(id_pool->ids);
+		id_pool->ids = NULL;
+		return ret;
+	}
+
+	id_pool->size = size;
+
+	return ret;
+}
+
+/*
+ * @brief	Allocate an ID
+ *
+ * Extract an available ID from the ID pool, mark the ID as
+ * unavailable in the pool, and return extracted ID. No-op in case
+ * no ID was available.
+ *
+ * This operation is locked by the ID pool's internal lock.
  *
- * @param	key_pool
- *		The Key pool
- * @return	Extracted key, on susccess
- *		FI_KEY_NOTAVAIL, in case no key is available
+ * @param	id_pool
+ *		The ID pool
+ * @param	id
+ *		Set to the extracted ID (zero-based), on success
+ * @return	0 on success
+ *		non-zero on error
  */
-uint64_t nccl_net_ofi_allocate_mr_key(nccl_ofi_mr_keypool_t *key_pool)
+int nccl_net_ofi_allocate_id(nccl_ofi_id_pool_t *id_pool, uint64_t *id)
 {
-	uint64_t key = FI_KEY_NOTAVAIL;
-	bool* mr_keys = key_pool->mr_keys;
-	int num_mr_keys = key_pool->size;
-	pthread_mutex_t *lock = &key_pool->lock;
+	if (OFI_UNLIKELY(NULL == id_pool || NULL == id)) {
+		NCCL_OFI_WARN("Invalid call to nccl_net_ofi_allocate_id with null pointer");
+		return -EINVAL;
+	}
 
-	if (mr_keys == NULL) {
-		NCCL_OFI_WARN("Invalid call to allocate_mr_key");
-		return FI_KEY_NOTAVAIL;
+	if (OFI_UNLIKELY(NULL == id_pool->ids)) {
+		NCCL_OFI_WARN("Invalid call to nccl_net_ofi_allocate_id with uninitialized pool");
+		return -EINVAL;
 	}
 
-	pthread_mutex_lock(lock);
+	pthread_mutex_lock(&id_pool->lock);
 
-	for (size_t i = 0; i < num_mr_keys; i++) {
-		if (mr_keys[i]) {
-			mr_keys[i] = false;
-			key = i;
+	int entry_index = 0;
+	for (size_t i = 0; i < id_pool->entries; i++) {
+		entry_index = __builtin_ffsll(id_pool->ids[i]);
+		if (0 != entry_index) {
+			/* Found one available ID */
+
+			/* Set to 0 bit at entry_index - 1 */
+			id_pool->ids[i] &= ~(1ULL << (entry_index - 1));
+
+			/* Store the ID we found */
+			*id = (i * sizeof(uint64_t) * 8) + entry_index - 1;
 			break;
 		}
 	}
 
-	if (key == FI_KEY_NOTAVAIL)
-		NCCL_OFI_WARN("No MR keys available (max: %d)", num_mr_keys);
+	pthread_mutex_unlock(&id_pool->lock);
 
-	pthread_mutex_unlock(lock);
-	return key;
+	if (0 == entry_index) {
+		NCCL_OFI_WARN("No IDs available (max: %lu)", id_pool->size);
+		return -ENOMEM;
+	}
+
+	return 0;
 }
 
 /*
- * @brief	Free a memory registration key
+ * @brief	Free an ID from the pool
  *
- * Return input key into the key pool.
+ * Return input ID into the pool.
  *
- * This operation is locked by the key pool's internal lock.
+ * This operation is locked by the ID pool's internal lock.
+ *
+ * @param	id_pool
+ *		The ID pool
+ * @param	id
+ *		The ID to release (zero-based)
+ * @return	0 on success
+ *		non-zero on error
  */
-int nccl_net_ofi_free_mr_key(nccl_ofi_mr_keypool_t *key_pool, uint64_t key)
+int nccl_net_ofi_free_id(nccl_ofi_id_pool_t *id_pool, uint64_t id)
```

Comment:
we're not consistent enough in the code, but please name <prefix>_<class>_<action>